Payment expiry for pending payments

ubbabeck opened the pull request:
This pr aims to add expiry time to pending payments.

Fixes [#679](https://github.com/breez/c-breez/issues/679)

## For reviews

I am new to rust, so if you see something which is not best practise or could be done functionally or in a _rustier_ way  please let me know.

roeierez commented on libs/sdk-core/src/breez_services.rs:
```diff
@@ -903,7 +903,29 @@ impl BreezServices {
             let closed_channel_tx = self.closed_channel_to_transaction(closed_channel).await?;
             closed_channel_payments.push(closed_channel_tx);
         }
+        let peers = self.node_api.list_peers().await?;
+
+        for p in peers {
+            for h in p.channels {
+                let htlcs = h.htlc;
+                if let Some(hh) = htlcs {
+                    for cc in hh {
+                        info!("htlc info {:?}{:?}", cc.payment_hash, cc.expiry);
+                        let payment_hash = hex::encode(cc.payment_hash);
+
+                        if let Ok(Some(payment)) = self.persister.get_payment_by_hash(&payment_hash)
+                        {
+                            info!(
+                                "htlc payment is {:?}",
+                                payment.status == PaymentStatus::Pending
+                            );
+                        }
+                    }
+                }
+            }
+        }
```
This whole logic of matching the htlc to the specific pending payment should be inside the NodeAPI implementation.

ubbabeck replied:
Done moved to NodeAPI

roeierez commented on libs/sdk-core/src/breez_services.rs:
```diff
@@ -910,8 +910,21 @@ impl BreezServices {
             closed_channel_payments.push(closed_channel_tx);
         }
 
+        let pending_payments: Vec<Payment> = new_data
+            .payments
+            .clone()
+            .into_iter()
+            .filter(|c| c.status == PaymentStatus::Pending)
+            .collect();
+        let mut payments: Vec<Payment> = vec![];
+        if !pending_payments.is_empty() {
+            info!("pending paymetns are not empty {:?}", pending_payments);
+            payments.extend(self.node_api.htlc_expiry(pending_payments).await?);
+        }
+        info!("payment list consist of {:?}", payments);
```
I think all this needs to be inside node_api. We don't need another method for `htlc_expiry` and we rather ensure the `pull_changed` method returns the payments with the right expiry populated.

ubbabeck replied:
Moved this logic.

ubbabeck commented:
Changes since last commit include converting block difference to unix time and persisting the payment_expiry to the db.

What needs to be fixed is:

- [ ] When a payment consists of multiple htlcs, we only need to care about the one expiring first
- [ ] Don't iterate over already pending payments more than once, atm it goes though this on every sync.

ubbabeck commented:
Instead of converting number of block to a unix timestamp it might be better to let return the block when the htlc is expiring. 

The reasoning is this : 

- even though the blocktime is approximately 10min it's usually quicker. 
- Developers might want to display block expiry rather than the calculated time, this calculation can be handled in the UI
- Relying on a timestamp timeout can backfire if notifications are sent too late. Even though we'll eventually add a webhook which handles this.
- The htlc will most likely be settled long before this timeout.

ubbabeck commented on libs/sdk-core/src/greenlight/node_api.rs:
```diff
@@ -1452,13 +1452,53 @@ async fn pull_transactions(
         .map(TryInto::try_into)
         .collect();
 
+    let res = c
```
True, fixed.

roeierez commented on libs/sdk-core/src/models.rs:
```diff
@@ -1098,6 +1109,23 @@ pub struct Channel {
     /// This may be empty for older closed channels, if it was not possible to retrieve the closing txid.
     pub closing_txid: Option<String>,
+
+    pub htlc: Option<Vec<Htlc>>,
```
I see this field is never None in the code so we can omit the Option as the vector can be empty instead.

ubbabeck replied:
Done

roeierez commented on libs/sdk-bindings/src/breez_sdk.udl:
```diff
@@ -282,6 +282,7 @@ dictionary LnPaymentDetails {
     string? ln_address;
     string? lnurl_withdraw_endpoint;
     SwapInfo? swap_info;
+    u32? payment_expiry;
```
Perhaps `pending_expiration_block` ?

ubbabeck replied:
Agree, renamed.

roeierez reviewed (approved):
LGTM

dangeross commented on libs/sdk-core/src/greenlight/node_api.rs:
```diff
@@ -1452,13 +1458,40 @@ async fn pull_transactions(
         .collect();
 
+    let outbound_transactions: NodeResult<Vec<Payment>> =
+        htlc_expiry(outbound_transactions.unwrap(), htlc_list);
```
To avoid a potential panic
```suggestion
        htlc_expiry(outbound_transactions?, htlc_list);
```

dangeross commented on libs/sdk-core/src/greenlight/node_api.rs:
```diff
@@ -1452,13 +1458,40 @@ async fn pull_transactions(
     Ok(transactions)
 }
 
+fn htlc_expiry(payments: Vec<Payment>, htlc_list: Vec<Htlc>) -> NodeResult<Vec<Payment>> {
```
The fn name doesn't really tell me what the function does, I had to look at the code. Maybe rename to `update_payment_expirations`

dangeross commented on libs/sdk-core/src/models.rs:
```diff
@@ -1098,6 +1109,23 @@ pub struct Channel {
     /// This may be empty for older closed channels, if it was not possible to retrieve the closing txid.
     pub closing_txid: Option<String>,
+
+    pub htlc: Vec<Htlc>,
```
I think this should be plural
```suggestion
    pub htlcs: Vec<Htlc>,
```

dangeross commented on libs/sdk-core/src/greenlight/node_api.rs:
```diff
@@ -1452,13 +1458,40 @@ async fn pull_transactions(
 }
 
+fn htlc_expiry(payments: Vec<Payment>, htlc_list: Vec<Htlc>) -> NodeResult<Vec<Payment>> {
+    let mut pending_payments: Vec<Payment> = Vec::new();
```
From what I understand in `pull_transactions` the payments could be pending or completed, maybe rename to something more general like `payments_res`. I assume if the payment is completed there will be no matching HTLC, so no need to filter.

dangeross commented on libs/sdk-core/src/models.rs:
```diff
@@ -1098,6 +1109,23 @@ pub struct Channel {
+    pub payment_hash: Vec<u8>,
+}
+
+impl Htlc {
+    pub fn from(expiry: u32, payment_hash: Vec<u8>) -> Self {
+        Htlc {
+            expiry,
+            payment_hash,
+        }
+    }
```
I don't think this is used, could be removed

ubbabeck replied:
I did not remove it and instead use it when listing channel peers. Not sure which approach is preferred.

dangeross reviewed (approved):
LGTM!
